File: tool/feedback/src/main/java/feedback/Algorithms.java
package feedback;

import feedback.diff.DistributedLogDiff;
import feedback.diff.LogDiff;
import feedback.diff.ThreadDiff;
import feedback.parser.DistributedLog;

import javax.json.JsonArray;
import javax.json.JsonObject;
import java.io.File;
import java.io.Serializable;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

final class Algorithms {
    private static void check(final DistributedLog good, final DistributedLog bad) throws Exception {
        if (good.distributed ^ bad.distributed) {
            throw new Exception("distributed mode not matched");
        }
        if (good.distributed &&
                !Arrays.equals(Arrays.stream(good.dirs).map(File::getName).toArray(String[]::new),
                        Arrays.stream(good.dirs).map(File::getName).toArray(String[]::new))) {
            throw new Exception("distributed log not matched");
        }
    }

    static void computeDiff(final DistributedLog good, final DistributedLog bad,
                            final Consumer<ThreadDiff.ThreadLogEntry> consumer) throws Exception {
        check(good, bad);
        if (good.distributed) {
            new DistributedLogDiff(good, bad).dumpBadDiff(consumer);
        } else {
            new LogDiff(good.logs[0], bad.logs[0]).dumpBadDiff(consumer);
        }
    }

    private static void computeLocationFeedback(final Consumer<Consumer<ThreadDiff.ThreadLogEntry>> dumpExpectedDiff,
                                                final Consumer<Consumer<ThreadDiff.ThreadLogEntry>> dumpActualDiff,
                                                final JsonObject spec, final Consumer<Integer> consumer)
            throws Exception {
        final int eventNumber = spec.getInt("start");
        final Map<ThreadDiff.ThreadLogEntry, Integer> wanted = new HashMap<>();
        dumpExpectedDiff.accept(e -> wanted.put(e, 0));
        if (wanted.size() + (Symptoms.isSymptomLogged(spec)? 0 : 1) != eventNumber) {
            throw new Exception("wrong diff");
        }
        final JsonArray array = spec.getJsonArray("nodes");
        for (int i = 0; i < array.size(); i++) {
            final JsonObject node = array.getJsonObject(i);
            final int id = node.getInt("id");
            if (id >= eventNumber) {
                continue;
            }
            if (id == 0) {
                continue;
            }
            if (!node.getString("type").equals("location_event")) {
                throw new Exception("wrong diff");
            }
            final JsonObject location = node.getJsonObject("location");
            final String cls = location.getString("class");
            final ThreadDiff.ThreadLogEntry entry =
                    new ThreadDiff.ThreadLogEntry(cls.substring(cls.lastIndexOf('.') + 1),
                            location.getInt("line_number"));
            if (wanted.containsKey(entry)) {
                wanted.put(entry, id);
            } else {
                throw new Exception("wrong diff");
            }
        }
        dumpActualDiff.accept(wanted::remove);
        if (Symptoms.isSymptomLogged(spec)) {
            wanted.values().forEach(i -> {
                if (i != 0) {
                    consumer.accept(i);
                }
            });
        } else {
            wanted.values().forEach(consumer);
        }
    }

    static void computeLocationFeedback(final DistributedLog good, final DistributedLog bad, final DistributedLog trial,
                                        final JsonObject spec, final Consumer<Integer> consumer) throws Exception {
        check(good, bad);
        check(good, trial);
        if (good.distributed) {
            computeLocationFeedback(new DistributedLogDiff(good, bad)::dumpBadDiff,
                    new DistributedLogDiff(good, trial)::dumpBadDiff, spec, consumer);
        } else {
            computeLocationFeedback(new LogDiff(good.logs[0], bad.logs[0])::dumpBadDiff,
                    new LogDiff(good.logs[0], trial.logs[0])::dumpBadDiff, spec, consumer);
        }
        if (!Symptoms.checkSymptom(trial, spec)) {
            consumer.accept(0);
        }
    }

//    static void computeTimeFeedback(final DistributedLog good, final DistributedLog bad, final DistributedLog trial,
//                                    final JsonObject spec, final Consumer<FeedbackTable> consumer) throws Exception {
//        check(good, bad);
//        check(good, trial);
//        if (good.distributed) {
//            for (int i = 0; i < good.logs.length; i++) {
//                final FeedbackTable table = computeTimeFeedback(good.logs[i], bad.logs[i], trial.logs[i], spec);
//                table.setPid(Parser.parseLogDirId(good.dirs[i].getName()));
//                consumer.accept(table);
//            }
//        } else {
//            computeTimeFeedback(good.logs[0], bad.logs[0], trial.logs[0], spec);
//        }
//    }
//

    static Serializable computeTimeFeedback(final DistributedLog good, final DistributedLog bad, final DistributedLog trial,
                                            final JsonObject spec) throws Exception {
        check(good, bad);
        check(good, trial);
    }
}

File: tool/feedback/src/main/scala/feedback/Algorithms.scala
package feedback

import feedback.common.ActionMayThrow
import feedback.diff.{DiffDump, LogFileDiff, ThreadDiff}
import feedback.log.{DistributedWorkloadLog, Log, UnitTestLog}
import feedback.parser.TextParser
import feedback.symptom.Symptoms
import feedback.time.TimeAlgorithms

import javax.json.JsonObject

object Algorithms {

  def computeLocationFeedback(good: Log, bad: Log, trial: Log, spec: JsonObject, action: ActionMayThrow[Integer]): Unit = {
    val expectedDiff = computeDiff(good, bad)
    val actualDiff = computeDiff(good, trial)
    val wanted = new java.util.HashMap[ThreadDiff.CodeLocation, Integer]
    expectedDiff foreach { _.dumpBadDiff { e => wanted.put(e, -1) } }
    val eventNumber = spec.getInt("start")
    require(eventNumber <= wanted.size + (if (Symptoms.isResultEventLogged(spec)) 0 else 1))
    val array = spec.getJsonArray("nodes")
    (0 until array.size) foreach { i =>
      val node = array.getJsonObject(i)
      val id = node.getInt("id")
      if (id < eventNumber && id != 0) {
        require(node.getString("type") equals "location_event")
        val location = node.getJsonObject("location")
        val entry = new ThreadDiff.CodeLocation(
          TextParser.getSimpleClassName(location.getString("class")), location.getInt("line_number"))
        require(wanted.put(entry, id) != null)
      }
    }
    actualDiff foreach { _.dumpBadDiff { e => wanted.remove(e) } }
    def filter(i: Int): Unit = if (i != -1) action.accept(i)
    if (Symptoms.isResultEventLogged(spec)) {
      wanted.values.forEach { i =>
        if (i != 0) {
          filter(i)
        }
      }
    } else wanted.values.forEach { filter(_) }
    if (Symptoms.findResultEvent(trial, spec).isEmpty) {
      filter(0)
    }
  }

  def computeTimeFeedback(good: Log, bad: Log, spec: JsonObject): Serializable = {
    val eventNumber = spec.getInt("start")
    val isResultEventLogged = Symptoms.isResultEventLogged(spec)
    val array = spec.getJsonArray("nodes")
    val events = Array.fill[Option[ThreadDiff.CodeLocation]](eventNumber)(None)
    (0 until array.size) foreach { i =>
      val node = array.getJsonObject(i)
      val id = node.getInt("id")
      if (id < eventNumber && (id != 0 || isResultEventLogged)) {
        require(node.getString("type") equals "location_event")
        val location = node.getJsonObject("location")
        val entry = new ThreadDiff.CodeLocation(
          TextParser.getSimpleClassName(location.getString("class")), location.getInt("line_number"))
        events(id) = Some(entry)
      }
    }
    TimeAlgorithms.computeTimeFeedback(good, bad, spec, events)
  }

  def computeDiff(good: Log, bad: Log): List[DiffDump] = (good, bad) match {
    case (UnitTestLog(goodLogFile, _), UnitTestLog(badLogFile, _)) =>
      List(new LogFileDiff(goodLogFile, badLogFile))
    case (DistributedWorkloadLog(goodLogs), DistributedWorkloadLog(badLogs)) =>
      require(goodLogs.length == badLogs.length)
      goodLogs.zip(badLogs).toList map {
        case (goodLogFile, badLogFile) =>
          new LogFileDiff(goodLogFile, badLogFile)
      }
  }

  def computeDoubleDiff1(good: Log, bad: Log, trial: Log, action: ActionMayThrow[ThreadDiff.CodeLocation]): Unit = {
    val set = new java.util.HashSet[ThreadDiff.CodeLocation]
    computeDiff(good, bad) foreach { _.dumpBadDiff(e =>
      set.add(e)
    )}
    computeDiff(good, trial) foreach { _.dumpBadDiff(e =>
      set.remove(e)
    )}
    set.forEach { i =>
      action.accept(i);
    }
  }

  def computeDoubleDiff(good: Log, bad: Log, trial: Log, action: ActionMayThrow[ThreadDiff.CodeLocation]): Unit = {
    val set = new java.util.HashSet[ThreadDiff.CodeLocation]
    val expected = computeDiff(good, bad)
    val actual = computeDiff(good, trial)
    val result = (actual zip expected).map{ case (a, e) =>
      new ThreadDiff(a.sortCodeLocationInThreadOrder(),e.sortCodeLocationInThreadOrder())}
    result foreach {
      _.dumpBadDiff(e =>
        set.add(e)
      )
    }
    set.forEach { i =>
      action.accept(i);
    }
  }

  def computeDiff(good: Log, bad: Log, action: ActionMayThrow[ThreadDiff.CodeLocation]): Unit = {
    val set = new java.util.HashSet[ThreadDiff.CodeLocation]
    computeDiff(good, bad) foreach { _.dumpBadDiff(e =>
      if (set.add(e)) {
        action.accept(e)
      }
    )}
  }
}
